src/SonarScanner.MSBuild.PreProcessor/ProcessedArgs.cs

```diff
@@ -151,5 +155,43 @@ public string GetSetting(string key, string defaultValue)
 
         public IEnumerable<Property> AllProperties() =>
             AggregateProperties.GetAllProperties();
+
+        // see spec in https://xtranet-sonarsource.atlassian.net/wiki/spaces/LANG/pages/3155001395/Scanner+Bootstrappers+implementation+guidelines
+        private static SonarServer GetSonarServer(ILogger logger, bool isHostSet, string sonarHostUrl, bool isSonarcloudSet, string sonarcloudUrl)
+        {
+            const string defaultSonarCloud = "https://sonarcloud.io";
+            if (isHostSet && isSonarcloudSet)
+            {
+                if (sonarHostUrl != sonarcloudUrl)
+                {
+                    logger.LogError(Resources.ERR_HostUrlDiffersFromSonarcloudUrl);
+                    return null;
+                }
+                else if (string.IsNullOrWhiteSpace(sonarcloudUrl))
+                {
+                    logger.LogError(Resources.ERR_HostUrlAndSonarcloudUrlAreEmpty);
+                    return null;
+                }
+                else
+                {
+                    logger.LogWarning(Resources.WARN_HostUrlAndSonarcloudUrlSet);
+                    return new SonarCloudServer(sonarcloudUrl);
+                }
+            }
+            else if (!isHostSet && !isSonarcloudSet)
+            {
+                return new SonarCloudServer(defaultSonarCloud);
+            }
+            else if (isHostSet) // isSonarcloudSet is false here
+            {
+                return sonarHostUrl == defaultSonarCloud
+                    ? new SonarCloudServer(defaultSonarCloud)
+                    : new SonarQubeServer(sonarHostUrl);
+            }
+            else // isHostSet is false and isSonarcloudSet is true here
+            {
+                return new SonarCloudServer(sonarcloudUrl);
+            }
+        }
```

Comment:
What do you think about this approach with switch expressions:
```csharp
                   return new { isHostSet, isSonarcloudSet} switch
                   {
                       { isHostSet: true, isSonarcloudSet: true } when sonarHostUrl != sonarcloudUrl => Error(Resources.ERR_HostUrlDiffersFromSonarcloudUrl),
                       { isHostSet: true, isSonarcloudSet: true } when string.IsNullOrWhiteSpace(sonarcloudUrl) => Error(Resources.ERR_HostUrlAndSonarcloudUrlAreEmpty),
                       { isHostSet: true, isSonarcloudSet: true } => new SonarCloudServer(sonarcloudUrl),
                   };

           SonarServer Error(string message)
           {
               logger.LogError(message);
               return null;
           }
```
I personally find it much easier to read and understand. It's incomplete though so you'll have to add the missing branches.

Reply:
I tried with tuples but didn't think about anonymous types because they do not support deconstruction. But for pattern matching deconstruction isn't needed 🤦 .